File: apps/api-py/eq_api/routes/projects.py
"""
Proyectos: el CRUD, con la misma tarjeta que pinta la lista del front.

Toda ruta cuelga de `/orgs/{organizationId}/`, así que el límite entre inquilinos lo pone el mismo
rol de siempre sobre un valor que está en la URL. El identificador del proyecto se comprueba
**otra vez** dentro, contra esa organización: el rol prueba que perteneces a la *organización*, y
solo la comprobación de dentro prueba que el *proyecto* también es de ella.

Roles, como en el original: `viewer` lee, `editor` crea y edita, `admin` archiva y borra.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Request
from pydantic import Field

from ..db import iso
from ..domain import base_url_problems, normalize_tags, slugify_project, utc_now, view_project_auth
from ..identity import Principal, authenticate, role_in_organization
from ..problems import ConflictError, InvalidInputError, NotFoundError
from ..repositories import new_id
from ..validation import Rules
from .auth import StrictModel

router = APIRouter(prefix="/orgs/{organization_id}/projects", tags=["projects"])


class CreateProjectBody(StrictModel):
    name: str
    description: str | None = None
    baseUrl: str | None = None
    tags: list[str] | None = Field(default=None, max_length=30)


class UpdateProjectBody(StrictModel):
    name: str | None = None
    description: str | None = None
    baseUrl: str | None = None
    tags: list[str] | None = Field(default=None, max_length=30)


class ArchiveProjectBody(StrictModel):
    archived: bool


def actor_id(principal: Principal) -> str:
    """Quien actúa. Un token de CI es un importador legítimo —así es como una tubería mantiene
    fresco el contrato— y por eso esto no exige una sesión de persona."""
    return principal.user_id or principal.token_id or ""


async def summarize(request: Request, project: dict) -> dict:
    """La tarjeta de un proyecto: lo suyo, más lo que solo se sabe mirando alrededor.

    `contract: null` es un estado real y la interfaz lo pinta: un proyecto existe antes de su
    primera importación, porque importar puede fallar y perder el proyecto con ello no ayuda."""
    read = request.app.state.eq.repositories.project_read
    projects = request.app.state.eq.repositories.projects

    last_run = await read.last_run(str(project["id"]))
    active = await read.active_spec_version(str(project["activeSpecVersionId"]) if project["activeSpecVersionId"] else None)
    source = await read.latest_source(str(project["id"]))
    fork = await read.fork(str(project["id"]))
    parent = await projects.find_by_id(str(fork["parentProjectId"])) if fork else None

    return {
        "id": str(project["id"]),
        "name": project["name"],
        "slug": project["slug"],
        "description": project["description"],
        "archivedAt": iso(project["archivedAt"]),
        "baseUrl": project["baseUrl"],
        "activeEnvironmentId": str(project["activeEnvironmentId"]) if project["activeEnvironmentId"] else None,
        "tags": project["tags"],
        "auth": view_project_auth(project["authType"], project["authSettings"] or {}),
        "lastRun": (
            {
                "id": str(last_run["id"]),
                "status": last_run["status"],
                "startedAt": iso(last_run["startedAt"]),
                "finishedAt": iso(last_run["finishedAt"]),
                "totals": last_run["totals"],
            }
            if last_run
            else None
        ),
        "contract": (
            {
                "versionId": str(active["id"]),
                "title": active["title"],
                "version": active["contractVersion"],
                "operationCount": active["operationCount"],
                "importedAt": iso(active["importedAt"]),
            }
            if active
            else None
        ),
        "source": (
            {
                "kind": source["kind"],
                "location": source["location"],
                "headersStored": source["headersCiphertext"] is not None,
            }
            if source
            else None
        ),
        "fork": (
            {
                "parentProjectId": str(fork["parentProjectId"]),
                # El nombre del original solo si sigue estando y es de esta organización: uno
                # borrado no se nombra.
                "parentName": (
                    parent["name"]
                    if parent and str(parent["organizationId"]) == str(project["organizationId"])
                    else None
                ),
                "forkedAt": iso(fork["createdAt"]),
                "syncedAt": iso(fork["syncedAt"]),
                "version": fork["version"],
            }
            if fork
            else None
        ),
    }


async def owned_project(request: Request, organization_id: str, project_id: str) -> dict:
    project = await request.app.state.eq.repositories.projects.find_by_id(project_id)
    if not project or str(project["organizationId"]) != organization_id:
        raise NotFoundError("El proyecto no existe", "project-not-found")
    return project


@router.get("")
async def list_projects(
    request: Request,
    organization_id: str,
    includeArchived: str | None = None,  # noqa: N803 — el nombre del parámetro es parte del contrato
    principal: Principal = Depends(authenticate),
) -> list:
    await role_in_organization(request, principal, organization_id, "viewer")
    projects = await request.app.state.eq.repositories.projects.list_for_organization(
        organization_id, includeArchived == "true"
    )
    return [await summarize(request, project) for project in projects]


@router.post("", status_code=201)
async def create_project(
    request: Request,
    organization_id: str,
    body: CreateProjectBody,
    principal: Principal = Depends(authenticate),
) -> dict:
    await role_in_organization(request, principal, organization_id, "editor")
    Rules().max_length("name", body.name, 200).max_length("description", body.description, 2000).max_length(
        "baseUrl", body.baseUrl, 2000
    ).check()
    state = request.app.state.eq

    problems = base_url_problems(body.baseUrl)
    if problems:
        raise InvalidInputError("La configuración del proyecto no es válida", problems)

    base = slugify_project(body.name)
    slug = base
    suffix = 2
    while await state.repositories.projects.find_by_slug(organization_id, slug):
        slug = f"{base}-{suffix}"
        suffix += 1
        if suffix >= 1000:
            slug = f"{base}-{new_id()[:8]}"
            break

    project_id = new_id()
    await state.repositories.projects.insert(
        {
            "id": project_id,
            "organizationId": organization_id,
            "name": body.name.strip(),
            "slug": slug,
            "description": (body.description or "").strip(),
            "createdBy": actor_id(principal),
            "createdAt": utc_now(),
            "baseUrl": (body.baseUrl or "").strip(),
            "tags": normalize_tags(body.tags or []),
        }
    )
    return {"projectId": project_id, "slug": slug}


@router.get("/{project_id}")
async def get_project(
    request: Request,
    organization_id: str,
    project_id: str,
    principal: Principal = Depends(authenticate),
) -> dict:
    await role_in_organization(request, principal, organization_id, "viewer")
    return await summarize(request, await owned_project(request, organization_id, project_id))


@router.patch("/{project_id}", status_code=204)
async def update_project(
    request: Request,
    organization_id: str,
    project_id: str,
    body: UpdateProjectBody,
    principal: Principal = Depends(authenticate),
) -> None:
    await role_in_organization(request, principal, organization_id, "editor")
    Rules().max_length("name", body.name, 200).max_length("description", body.description, 2000).max_length(
        "baseUrl", body.baseUrl, 2000
    ).check()
    project = await owned_project(request, organization_id, project_id)
    if project["archivedAt"] is not None:
        raise ConflictError("El proyecto está archivado", "project-archived")

    problems = base_url_problems(body.baseUrl)
    if problems:
        raise InvalidInputError("La configuración del proyecto no es válida", problems)

    # El slug **no** se recalcula desde un nombre nuevo: está en URLs que el equipo tiene
    # guardadas y en el trabajo de CI que lanza sus corridas. Renombrar no rompe ninguna de las dos.
    await request.app.state.eq.repositories.projects.update_fields(
        project_id,
        name=(body.name or "").strip() or project["name"],
        description=(body.description.strip() if body.description is not None else project["description"]),
        base_url=(body.baseUrl.strip() if body.baseUrl is not None else project["baseUrl"]),
        tags=(normalize_tags(body.tags) if body.tags is not None else project["tags"]),
    )


@router.patch("/{project_id}/archived", status_code=204)
async def set_archived(
    request: Request,
    organization_id: str,
    project_id: str,
    body: ArchiveProjectBody,
    principal: Principal = Depends(authenticate),
) -> None:
    # Archivar saca el proyecto de la lista de todo el mundo, así que sube un escalón sobre editarlo.
    await role_in_organization(request, principal, organization_id, "admin")
    await owned_project(request, organization_id, project_id)
    await request.app.state.eq.repositories.projects.set_archived(
        project_id, utc_now() if body.archived else None
    )


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    request: Request,
    organization_id: str,
    project_id: str,
    principal: Principal = Depends(authenticate),
) -> None:
    # Borrar es definitivo para toda la organización, así que es el mismo escalón que archivar.
    await role_in_organization(request, principal, organization_id, "admin")
    await owned_project(request, organization_id, project_id)
    await request.app.state.eq.repositories.projects.soft_delete(project_id, utc_now())

File: apps/api-go/internal/api/projects.go
package api

// Proyectos: el CRUD, con la misma tarjeta que pinta la lista del front.
//
// Toda ruta cuelga de `/orgs/{organizationId}/`, así que el límite entre inquilinos lo pone el rol
// sobre un valor que está en la URL. El identificador del proyecto se comprueba **otra vez**
// dentro, contra esa organización: el rol prueba que perteneces a la *organización*, y solo la
// comprobación de dentro prueba que el *proyecto* también es de ella.

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/smalbach/endpoint-quality/apps/api-go/internal/domain"
	"github.com/smalbach/endpoint-quality/apps/api-go/internal/problems"
	"github.com/smalbach/endpoint-quality/apps/api-go/internal/store"
)

type createProjectBody struct {
	Name        string   `json:"name"`
	Description *string  `json:"description"`
	BaseURL     *string  `json:"baseUrl"`
	Tags        []string `json:"tags"`
}

type updateProjectBody struct {
	Name        *string  `json:"name"`
	Description *string  `json:"description"`
	BaseURL     *string  `json:"baseUrl"`
	Tags        []string `json:"tags"`
}

type archiveProjectBody struct {
	Archived bool `json:"archived"`
}

type projectCreated struct {
	ProjectID string `json:"projectId"`
	Slug      string `json:"slug"`
}

// La mitad no secreta de la autenticación de un proyecto, con una máscara donde hay un secreto
// guardado. El texto cifrado no sale de la base de datos por ninguna ruta.
type projectAuthView struct {
	Type        string `json:"type"`
	LoginURL    string `json:"loginUrl"`
	LoginMethod string `json:"loginMethod"`
	TokenPath   string `json:"tokenPath"`
	Username    string `json:"username"`
	HeaderName  string `json:"headerName"`
	Token       string `json:"token"`
	LoginBody   string `json:"loginBody"`
	Password    string `json:"password"`
	APIKey      string `json:"apiKey"`
}

type lastRunView struct {
	ID         string  `json:"id"`
	Status     string  `json:"status"`
	StartedAt  string  `json:"startedAt"`
	FinishedAt *string `json:"finishedAt"`
	Totals     any     `json:"totals"`
}

type contractView struct {
	VersionID      string `json:"versionId"`
	Title          string `json:"title"`
	Version        string `json:"version"`
	OperationCount int    `json:"operationCount"`
	ImportedAt     string `json:"importedAt"`
}

type sourceView struct {
	Kind          string `json:"kind"`
	Location      string `json:"location"`
	HeadersStored bool   `json:"headersStored"`
}

type forkView struct {
	ParentProjectID string  `json:"parentProjectId"`
	ParentName      *string `json:"parentName"`
	ForkedAt        string  `json:"forkedAt"`
	SyncedAt        string  `json:"syncedAt"`
	Version         int     `json:"version"`
}

type projectSummary struct {
	ID                  string          `json:"id"`
	Name                string          `json:"name"`
	Slug                string          `json:"slug"`
	Description         string          `json:"description"`
	ArchivedAt          *string         `json:"archivedAt"`
	BaseURL             string          `json:"baseUrl"`
	ActiveEnvironmentID *string         `json:"activeEnvironmentId"`
	Tags                []string        `json:"tags"`
	Auth                projectAuthView `json:"auth"`
	LastRun             *lastRunView    `json:"lastRun"`
	Contract            *contractView   `json:"contract"`
	Source              *sourceView     `json:"source"`
	Fork                *forkView       `json:"fork"`
}

const mask = "••••••••"

func viewProjectAuth(authType string, settings map[string]any) projectAuthView {
	text := func(key string) string {
		if value, ok := settings[key].(string); ok {
			return value
		}
		return ""
	}
	secrets := map[string]bool{}
	if list, ok := settings["secretFields"].([]any); ok {
		for _, field := range list {
			if name, ok := field.(string); ok {
				secrets[name] = true
			}
		}
	}
	masked := func(field string) string {
		if secrets[field] {
			return mask
		}
		return ""
	}
	return projectAuthView{
		Type:        authType,
		LoginURL:    text("loginUrl"),
		LoginMethod: text("loginMethod"),
		TokenPath:   text("tokenPath"),
		Username:    text("username"),
		HeaderName:  text("headerName"),
		Token:       masked("token"),
		LoginBody:   masked("loginBody"),
		Password:    masked("password"),
		APIKey:      masked("apiKey"),
	}
}

// summarize arma la tarjeta: lo del proyecto, más lo que solo se sabe mirando alrededor.
// `contract: null` es un estado real y la interfaz lo pinta — un proyecto existe antes de su
// primera importación, porque importar puede fallar y perder el proyecto con ello no ayuda.
func (s *Server) summarize(request *http.Request, project store.Project) (projectSummary, error) {
	ctx := request.Context()
	summary := projectSummary{
		ID:                  project.ID,
		Name:                project.Name,
		Slug:                project.Slug,
		Description:         project.Description,
		ArchivedAt:          isoPointer(project.ArchivedAt),
		BaseURL:             project.BaseURL,
		ActiveEnvironmentID: project.ActiveEnvironmentID,
		Tags:                project.Tags,
		Auth:                viewProjectAuth(project.AuthType, project.AuthSettings),
	}
	if summary.Tags == nil {
		summary.Tags = []string{}
	}

	run, err := s.store.LastRun(ctx, project.ID)
	if err != nil {
		return summary, err
	}
	if run != nil {
		summary.LastRun = &lastRunView{
			ID: run.ID, Status: run.Status, StartedAt: iso(run.StartedAt),
			FinishedAt: isoPointer(run.FinishedAt), Totals: run.Totals,
		}
	}

	if project.ActiveSpecVersionID != nil {
		version, err := s.store.SpecVersion(ctx, *project.ActiveSpecVersionID)
		if err != nil {
			return summary, err
		}
		if version != nil {
			summary.Contract = &contractView{
				VersionID: version.ID, Title: version.Title, Version: version.ContractVersion,
				OperationCount: version.OperationCount, ImportedAt: iso(version.ImportedAt),
			}
		}
	}

	source, err := s.store.LatestSpecSource(ctx, project.ID)
	if err != nil {
		return summary, err
	}
	if source != nil {
		summary.Source = &sourceView{Kind: source.Kind, Location: source.Location, HeadersStored: source.HeadersStored}
	}

	fork, err := s.store.Fork(ctx, project.ID)
	if err != nil {
		return summary, err
	}
	if fork != nil {
		view := &forkView{
			ParentProjectID: fork.ParentProjectID, ForkedAt: iso(fork.CreatedAt),
			SyncedAt: iso(fork.SyncedAt), Version: fork.Version,
		}
		parent, err := s.store.ProjectByID(ctx, fork.ParentProjectID)
		if err != nil {
			return summary, err
		}
		// El nombre del original solo si sigue estando y es de esta organización: uno borrado no
		// se nombra.
		if parent != nil && parent.OrganizationID == project.OrganizationID {
			name := parent.Name
			view.ParentName = &name
		}
		summary.Fork = view
	}
	return summary, nil
}

func (s *Server) ownedProject(request *http.Request, organizationID, projectID string) (*store.Project, error) {
	project, err := s.store.ProjectByID(request.Context(), projectID)
	if err != nil {
		return nil, err
	}
	if project == nil || project.OrganizationID != organizationID {
		return nil, problems.NotFound("El proyecto no existe", "project-not-found")
	}
	return project, nil
}

func (s *Server) listProjects(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "viewer"); err != nil {
		return err
	}
	projects, err := s.store.Projects(request.Context(), organizationID,
		request.URL.Query().Get("includeArchived") == "true")
	if err != nil {
		return err
	}
	summaries := []projectSummary{}
	for _, project := range projects {
		summary, err := s.summarize(request, project)
		if err != nil {
			return err
		}
		summaries = append(summaries, summary)
	}
	return writeJSON(writer, http.StatusOK, summaries)
}

func (s *Server) createProject(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "editor"); err != nil {
		return err
	}
	var body createProjectBody
	if err := decode(request, &body); err != nil {
		return err
	}
	check := (rules{}).maxLength("name", body.Name, 200)
	if body.Description != nil {
		check = check.maxLength("description", *body.Description, 2000)
	}
	if body.BaseURL != nil {
		check = check.maxLength("baseUrl", *body.BaseURL, 2000)
	}
	if err := check.check(); err != nil {
		return err
	}
	if found := domain.BaseURLProblems(body.BaseURL); len(found) > 0 {
		return problems.Invalid("La configuración del proyecto no es válida", found, "")
	}

	ctx := request.Context()
	base := domain.SlugifyProject(body.Name)
	slug := base
	for suffix := 2; suffix < 1000; suffix++ {
		taken, err := s.store.ProjectSlugTaken(ctx, organizationID, slug)
		if err != nil {
			return err
		}
		if !taken {
			break
		}
		slug = fmt.Sprintf("%s-%d", base, suffix)
	}

	project := store.Project{
		ID: newID(), OrganizationID: organizationID, Name: strings.TrimSpace(body.Name), Slug: slug,
		Description: strings.TrimSpace(valueOr(body.Description, "")), CreatedAt: time.Now().UTC(),
		BaseURL: strings.TrimSpace(valueOr(body.BaseURL, "")), Tags: domain.NormalizeTags(body.Tags),
	}
	// Quien actúa: un token de CI es un importador legítimo, así que esto no exige una persona.
	actor := who.userID
	if actor == "" {
		actor = who.tokenID
	}
	if err := s.store.InsertProject(ctx, project, actor); err != nil {
		return err
	}
	return writeJSON(writer, http.StatusCreated, projectCreated{ProjectID: project.ID, Slug: slug})
}

func (s *Server) getProject(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "viewer"); err != nil {
		return err
	}
	project, err := s.ownedProject(request, organizationID, request.PathValue("projectId"))
	if err != nil {
		return err
	}
	summary, err := s.summarize(request, *project)
	if err != nil {
		return err
	}
	return writeJSON(writer, http.StatusOK, summary)
}

func (s *Server) updateProject(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	if _, err := s.roleIn(request, who, organizationID, "editor"); err != nil {
		return err
	}
	var body updateProjectBody
	if err := decode(request, &body); err != nil {
		return err
	}
	check := rules{}
	if body.Name != nil {
		check = check.maxLength("name", *body.Name, 200)
	}
	if body.Description != nil {
		check = check.maxLength("description", *body.Description, 2000)
	}
	if body.BaseURL != nil {
		check = check.maxLength("baseUrl", *body.BaseURL, 2000)
	}
	if err := check.check(); err != nil {
		return err
	}

	project, err := s.ownedProject(request, organizationID, request.PathValue("projectId"))
	if err != nil {
		return err
	}
	if project.ArchivedAt != nil {
		return problems.Conflict("El proyecto está archivado", "project-archived")
	}
	if found := domain.BaseURLProblems(body.BaseURL); len(found) > 0 {
		return problems.Invalid("La configuración del proyecto no es válida", found, "")
	}

	name := project.Name
	if body.Name != nil && strings.TrimSpace(*body.Name) != "" {
		name = strings.TrimSpace(*body.Name)
	}
	description := project.Description
	if body.Description != nil {
		description = strings.TrimSpace(*body.Description)
	}
	baseURL := project.BaseURL
	if body.BaseURL != nil {
		baseURL = strings.TrimSpace(*body.BaseURL)
	}
	tags := project.Tags
	if body.Tags != nil {
		tags = domain.NormalizeTags(body.Tags)
	}

	// El slug **no** se recalcula desde un nombre nuevo: está en URLs que el equipo tiene guardadas
	// y en el trabajo de CI que lanza sus corridas. Renombrar no rompe ninguna de las dos.
	if err := s.store.UpdateProject(request.Context(), project.ID, name, description, baseURL, tags); err != nil {
		return err
	}
	return noContent(writer)
}

func (s *Server) setProjectArchived(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	// Archivar saca el proyecto de la lista de todo el mundo, así que sube un escalón sobre editarlo.
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	var body archiveProjectBody
	if err := decode(request, &body); err != nil {
		return err
	}
	project, err := s.ownedProject(request, organizationID, request.PathValue("projectId"))
	if err != nil {
		return err
	}
	var at *time.Time
	if body.Archived {
		now := time.Now().UTC()
		at = &now
	}
	if err := s.store.SetProjectArchived(request.Context(), project.ID, at); err != nil {
		return err
	}
	return noContent(writer)
}

func (s *Server) deleteProject(writer http.ResponseWriter, request *http.Request) error {
	who, err := s.authenticate(request)
	if err != nil {
		return err
	}
	organizationID := request.PathValue("organizationId")
	// Borrar es definitivo para toda la organización, así que es el mismo escalón que archivar.
	if _, err := s.roleIn(request, who, organizationID, "admin"); err != nil {
		return err
	}
	project, err := s.ownedProject(request, organizationID, request.PathValue("projectId"))
	if err != nil {
		return err
	}
	if err := s.store.SoftDeleteProject(request.Context(), project.ID, time.Now().UTC()); err != nil {
		return err
	}
	return noContent(writer)
}

func valueOr(value *string, fallback string) string {
	if value == nil {
		return fallback
	}
	return *value
}
